### DISCUSSION TITLE
求教，如何找到chromium的二进制 ELF 文件路径

### DISCUSSION THREAD

**Discussion by Nigel-NI** [Arondight/Adachi-BOT#464] [Q&A] [answered]
我是debian 10系统，`epel-release`这个包没有找到，但chromium可以直接安装。
使用whereis查询到五个安装目录：

```
whereis chromium
chromium: /usr/bin/chromium /usr/lib/chromium /etc/chromium /etc/chromium.d /usr/share/chromium /usr/share/man/man1/chromium.1.gz

```
但我不清楚哪个路径是二进制 ELF 文件路径，还请指教，谢谢！

**Comment by Nigel-NI** [comment, marked as answer]
顺便请教一下，这个报错是不是因为二进制 ELF 文件路径错误导致的：
`genshin-package 功能绘图失败：Error: Tried to use PUPPETEER_EXECUTABLE_PATH env variable to launch browser but did not find any executable at: /usr/bin/chromium/chromium`

**Reply by Arondight** [reply]
效率低但是通用的方法

```
$ type chromium-browser 
chromium-browser is /bin/chromium-browser

$ file /bin/chromium-browser 
/bin/chromium-browser: symbolic link to /usr/lib64/chromium-browser/chromium-browser.sh

$ file /usr/lib64/chromium-browser/chromium-browser.sh                     
/usr/lib64/chromium-browser/chromium-browser.sh: Bourne-Again shell script, ASCII text executable

$ bash -x /usr/lib64/chromium-browser/chromium-browser.sh  
++ readlink -f /usr/lib64/chromium-browser/chromium-browser.sh
+ export CHROME_WRAPPER=/usr/lib64/chromium-browser/chromium-browser.sh
+ CHROME_WRAPPER=/usr/lib64/chromium-browser/chromium-browser.sh
++ dirname /usr/lib64/chromium-browser/chromium-browser.sh
+ HERE=/usr/lib64/chromium-browser
+ which xdg-settings
+ export PATH=/usr/lib64/chromium-browser:/home/user/bin:/home/user/lib/go/bin:/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin:.
+ PATH=/usr/lib64/chromium-browser:/home/user/bin:/home/user/lib/go/bin:/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin:.
+ [[ -n '' ]]
+ LD_LIBRARY_PATH=/usr/lib64/chromium-browser:/usr/lib64/chromium-browser/lib
+ export LD_LIBRARY_PATH
+ export 'CHROME_VERSION_EXTRA=Built from source for Red Hat Enterprise Linux release 8.4 (Ootpa)'
+ CHROME_VERSION_EXTRA='Built from source for Red Hat Enterprise Linux release 8.4 (Ootpa)'
+ export GNOME_DISABLE_CRASH_DIALOG=SET_BY_GOOGLE_CHROME
+ GNOME_DISABLE_CRASH_DIALOG=SET_BY_GOOGLE_CHROME
+ export allow_rgb10_configs=false
+ allow_rgb10_configs=false
+ CHROMIUM_DISTRO_FLAGS=' --enable-plugins                         --enable-extensions                         --enable-user-scripts                         --enable-printing                         --enable-gpu-rasterization                         --enable-sync                         --auto-ssl-client-auth '
+ '[' -f /usr/lib64/chromium-browser/PepperFlash/libpepflashplayer.so ']'
+ CHROMIUM_FLASH_FLAGS=
+ exec -a /usr/lib64/chromium-browser/chromium-browser.sh /usr/lib64/chromium-browser/chromium-browser --enable-plugins --enable-extensions --enable-user-scripts --enable-printing --enable-gpu-rasterization --enable-sync --auto-ssl-client-auth
[1379717:1379717:1202/164229.009578:ERROR:browser_main_loop.cc(1400)] Unable to open X display.

$ file /usr/lib64/chromium-browser/chromium-browser                                               
/usr/lib64/chromium-browser/chromium-browser: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=7577bb7f2ca3102627af85c02c904775d4a77a2d, for GNU/Linux 3.2.0, not stripped              
```

得出路径为 `/usr/lib64/chromium-browser/chromium-browser` 。

**Reply by Nigel-NI** [reply]
感谢回复，已解决。真实路径为/usr/bin/chromium

**Reply by Arondight** [reply]
跟我找的怎么不太一样 #465 ？ 难道是我用的 ubuntu 14 太老了？

**Reply by Arondight** [reply]
哦，你用的是 debian ， #465 里面 debian 的是我猜的，手里没 debian ，我还以为和 raspbian 一样呢

**Reply by Nigel-NI** [reply]
我是这样筛出结果的
```
root@debian:~# type chromium
chromium is /usr/bin/chromium
root@debian:~# file /usr/bin/chromium
/usr/bin/chromium: POSIX shell script, ASCII text executable
root@debian:~# bash -x /usr/bin/chromium
+ APPNAME=chromium
+ GDB=/usr/bin/gdb
+ LIBDIR=/usr/lib/chromium
+ BUILD_DIST=10.9
+ nosse2='The hardware on this system lacks support for the sse2 instruction set.
The upstream chromium project no longer supports this configuration.
For more information, please read and possibly provide input to their
bug tracking system at http://crbug.com/348761.'
+ case `uname -m` in
++ uname -m
+ grep -q sse2 /proc/cpuinfo
+ for file in /etc/chromium.d/*
以下略
```

**Reply by Arondight** [reply]
我特意去 debian 的源里面解了个包，看来我写的是对的，你直接抄把

**Reply by Nigel-NI** [reply]
可以，不过我把路径改成`/usr/bin/chromium`，也能正常运行，然后我就没再去动它……

**Reply by 12123ads** [reply]
请问：
[2022-01-03T08:33:50.084] [ERROR] [iPad:2058721672] - render： genshin-card-8 功能绘图失败：Error: Tried to use PUPPETEER_EXECUTABLE_PATH env variable to launch browser but did not find any executable at: “/usr/lib/chromium/chromium” 1770627763
这个问题是怎么解决的呢
我是debian11

**Reply by mark9804** [reply]
`where chromium`

> 请问： [2022-01-03T08:33:50.084] [ERROR] [iPad:2058721672] - render： genshin-card-8 功能绘图失败：Error: Tried to use PUPPETEER_EXECUTABLE_PATH env variable to launch browser but did not find any executable at: “/usr/lib/chromium/chromium” 1770627763 这个问题是怎么解决的呢 我是debian11

**Reply by 12123ads** [reply]
root@debian:~/Adachi-BOT# where chromium
-bash: where: command not found

**Reply by mark9804** [reply]
那看看
`type chromium` 或者 `type chromium-browser`，这两个应该有一个是可以返回的，我手上 debian 系的只有一个 raspbian

**Reply by mark9804** [reply]
其实就是主楼的回复，你按照主楼的方法试试看

**Comment by 12123ads** [comment]
***@***.***:~/Adachi-BOT#&nbsp;type&nbsp;chromium 
chromium&nbsp;is&nbsp;/usr/bin/chromium
这个有


---原始邮件---
发件人: &quot;Mark ***@***.***&gt;
发送时间: 2022年1月3日(周一) 下午5:04
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [Arondight/Adachi-BOT] 求教，如何找到chromium的二进制 ELF 文件路径 (Discussion #464)




 
那看看
 type chromium 或者 type chromium-browser，这两个应该有一个是可以返回的，我手上 debian 系的只有一个 raspbian
 
&mdash;

**Reply by mark9804** [reply]
那应该和楼上回答里面的一样，也是 `/usr/bin/chromium`

https://github.com/Arondight/Adachi-BOT/discussions/464#discussioncomment-1735808
https://github.com/Arondight/Adachi-BOT/discussions/464#discussioncomment-1735692

**Comment by 12123ads** [comment]
[2022-01-03T09:09:21.640] [ERROR] [iPad:2058721672] - render： genshin-card-8 功能绘图失败：Error: Tried to use PUPPETEER_EXECUTABLE_PATH env variable to launch browser but did not find any executable at: “/usr/bin/chromium/” 1770627763
但是这个目录也报错

**Reply by 12123ads** [reply]
![](https://i.bmp.ovh/imgs/2022/01/7402787e0b86f64e.png)
这个命令也有返回

**Reply by mark9804** [reply]
> “/usr/bin/chromium/”

你写的是 `/usr/bin/chromium` ，还是 `/usr/bin/chromium/`？
最后有没有斜杠的区别

**Reply by 12123ads** [reply]
[2022-01-03T09:17:19.103] [ERROR] [iPad:2058721672] - render： genshin-card-8 功能绘图失败：Error: Tried to use PUPPETEER_EXECUTABLE_PATH env variable to launch browser but did not find any executable at: “/usr/bin/chromium” 1770627763

**Reply by 12123ads** [reply]
这俩都不行

**Reply by mark9804** [reply]
那我不了解了，你等一等真大佬来

**Reply by 12123ads** [reply]
好的谢谢

**Reply by Arondight** [reply]
#465

**Reply by 12123ads** [reply]
这个路径还是报错啊
root@debian:~/Adachi-BOT# type chromium 
chromium is /usr/bin/chromium
root@debian:~/Adachi-BOT# file /usr/bin/chromium
/usr/bin/chromium: POSIX shell script, ASCII text executable

[2022-01-03T11:16:50.381] [DEBUG] [iPad:2058721672] - render：已生成 genshin-card-8 功能的数据调试文件。
[2022-01-03T11:16:50.388] [ERROR] [iPad:2058721672] - render： genshin-card-8 功能绘图失败：Error: Tried to use PUPPETEER_EXECUTABLE_PATH env variable to launch browser but did not find any executable at: “/usr/bin/chromium” 1770627763
[2022-01-03T11:16:50.457] [INFO] [iPad:2058721672] - succeed to send: [Private(1770627763)] 绘图失败。

**Reply by Arondight** [reply]
点击链接查看内容 #465

**Reply by 12123ads** [reply]
[2022-01-03T11:35:35.743] [ERROR] [iPad:2058721672] - render： genshin-card-8 功能绘图失败：Error: Tried to use PUPPETEER_EXECUTABLE_PATH env variable to launch browser but did not find any executable at: “/usr/lib/chromium/chromium” 1770627763

**Reply by Arondight** [reply]
确认一下安装的软件包是否一致，`dpkg -S /usr/lib/chromium/chromium`

**Reply by 12123ads** [reply]
root@debian:/Adachi-BOT# dpkg -S /usr/lib/chromium/chromium
chromium: /usr/lib/chromium/chromium
root@debian:/Adachi-BOT#

**Reply by 12123ads** [reply]
请问大佬这个怎么办

**Reply by Arondight** [reply]
`file /usr/lib/chromium/chromium` 看一下，是不是安装了不同架构的包，手动运行一下看看是否报错非可执行文件

**Reply by 12123ads** [reply]
/usr/lib/chromium/chromium: /usr/lib/chromium/chromium: cannot execute binary file
嘶，这咋办

**Reply by 12123ads** [reply]
我是通过apt安装的

**Reply by Arondight** [reply]
应该是安装的软件包架构和系统的不一致，例如你在树莓派中运行了一个 debian ，然后配置了一个 x86 的软件源，并安装了里面的包

**Reply by 12123ads** [reply]
系统是x86_64的没错

**Reply by Arondight** [reply]
file 命令看一下，这个路径到底是什么

**Reply by 12123ads** [reply]
root@debian:/Adachi-BOT# type chromium
chromium is hashed (/usr/bin/chromium)
root@debian:/Adachi-BOT# file /usr/bin/chromium
/usr/bin/chromium: POSIX shell script, ASCII text executable
root@debian:/Adachi-BOT# file /usr/lib/chromium/chromium
/usr/lib/chromium/chromium: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 3.2.0, BuildID[xxHash]=9ddeabb407fd4e5d, stripped
root@debian:/Adachi-BOT# 
应该是这个吧

**Reply by 12123ads** [reply]
root@debian:~/Adachi-BOT# chromium --no-sandbox --headless
[0103/124909.097372:WARNING:headless_browser_main_parts.cc(83)] Cannot create Pref Service with no user data dir.
[0103/124909.133983:ERROR:viz_main_impl.cc(150)] Exiting GPU process due to errors during initialization
[0103/124909.161739:WARNING:vaapi_wrapper.cc(588)] VAAPI video acceleration not available for disabled
[0103/124909.173595:ERROR:gpu_init.cc(426)] Passthrough is not supported, GL is disabled
这样的话是不是不能用？

**Reply by Arondight** [reply]
不能用，浏览器没起来

**Reply by 12123ads** [reply]
root@debian:~/Adachi-BOT# arch
x86_64
系统是86_64的

**Reply by Arondight** [reply]
这个得你自己查了，我搜了下似乎是有一些人遇到这个问题的，应该有很多可供参考的信息

**Reply by 12123ads** [reply]
![](https://i.bmp.ovh/imgs/2022/01/af53ef7f71c7d498.png)换了个机子，这为啥没反应

**Reply by Arondight** [reply]
大概是太卡了，chrome 跑不动了，`free -h` 看一下，swap 如果没有，建议挂个 swap 文件

**Reply by 12123ads** [reply]
![](https://i.bmp.ovh/imgs/2022/01/a4137eb1cbc55574.png)
一个G的swap不够吗

**Reply by Arondight** [reply]
你截图的 `956 MB` 是内存，swap 是 swap，建议分点看看

**Reply by 12123ads** [reply]
![](https://i.bmp.ovh/imgs/2022/01/d74481be59b77512.png)
没事了，好像是cpu不够了。。。

**Reply by 12123ads** [reply]
0|bot      | [2022-01-04T18:06:41.912] [DEBUG] [iPad:2058721672] - render：已生成 genshin-card-8 功能的数据调试文件。
0|bot      | [2022-01-04T18:08:56.311] [ERROR] [iPad:2058721672] - render： genshin-card-8 功能绘图失败：TimeoutError: Navigation timeout of 30000 ms exceeded 1770627763
请问这个问题是？

**Reply by Arondight** [reply]
内存不够了

**Reply by 12123ads** [reply]
[2022-01-07T07:52:47.202] [DEBUG] [iPad:2058721672] - render：已生成 genshin-card-8 功能的数据调试文件。
[2022-01-07T07:52:47.213] [ERROR] [iPad:2058721672] - render： genshin-card-8 功能绘图失败：Error: Failed to launch the browser process! spawn /usr/bin/chromium-browser ENOENT


TROUBLESHOOTING: https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md
 1770627763
这个是为啥

**Reply by Arondight** [reply]
优质答案：我不知道
